tests/30rooms/12thirdpartyinvite.pl

```diff
@@ -0,0 +1,419 @@
+#                     key_validity_url => "https://$id_server/_matrix/identity/api/v1/pubkey/isvalid",
+#                     sender => $user->user_id,
+#                  }
+#               )
+#               ->followed_by(\&main::expect_http_4xx)
+#               ->then( sub {
+#                  $do_request_json->(
+#                     method => "GET",
+#                     uri    => "/api/v1/rooms/$room_id/state/m.room.member/".$invitee->user_id,
+#                  )->followed_by(assert_membership( $do_request_json, $inviter, undef ) )
+#               })
+#            })
+#         }),
+#      );
+#   };
+
+# In order:
+#  1. Creates all state needed to issue a 3pid invite
+#  2. Issues the invite
+#  3. Calls join_sub with the following args: token (str), public_key (base64 str), $signature (base64 str), room_id (str)
+#  4. Asserts that invitee did/didn't join the room, depending on truthiness of expect_join_success
+#  5. Awaits on all passed futures, so that you can stub/mock things as you wish
+sub make_3pid_invite {
+   my ( $inviter, $invitee, $id_server, $expect_join_success, $join_sub, $futures ) = @_;
+
+   my $invitee_email = 'lemurs@monkeyworld.org';
+
+   my $token = "abc123";
+
+   my ( $public_key, $private_key ) = $crypto_sign->keypair;
+   my $encoded_public_key = encode_base64_unpadded( $public_key );
+   my $signature = encode_base64_unpadded( $crypto_sign->mac( $token, $private_key ) );
+
+   my $response_verifier = $expect_join_success
+      ? sub {
+         my ( $f ) = @_;
+         $f->then( sub {
+            Future->done( @_ );
+         }, sub {
+            my ( undef, $name, $response ) = @_;
+
+            Future->fail( @_ );
+         });
+      }
+      : \&main::expect_http_4xx;
+
+   my $room_id;
+
+   Future->needs_all(
+      stub_is_lookup( $invitee_email, undef ),
+
+      stub_is_token_generation( $token, $encoded_public_key ),
+
+      @$futures,
+
+      matrix_create_room( $inviter, visibility => "private" )
+      ->then(sub {
+         ( $room_id ) = @_;
+         do_3pid_invite( $inviter, $room_id, $id_server, $invitee_email )
+      })->then( sub {
+         $join_sub->( $token, $public_key, $signature, $room_id )
+      })->followed_by($response_verifier)
+      ->then( sub {
+         matrix_get_room_state( $inviter, $room_id,
+            type => "m.room.member",
+            state_key => $invitee->user_id,
+         )
+      })->followed_by(assert_membership( $inviter, $expect_join_success ? "join" : undef ) ),
+   );
+}
+
+sub assert_membership {
+   my ( $user, $expected_membership ) = @_;
+
+   my $verifier = defined $expected_membership
+      ? sub {
+         my ( $f ) = @_;
+
+         $f->then( sub {
+            my ( $body ) = @_;
+
+            log_if_fail "Body", $body;
+            $body->{membership} eq $expected_membership or
+               die "Expected invited user membership to be '$expected_membership' but was '".$body->membership."'";
```

Comment:
There's no need quoting the actual membership found in the body (and in any case you're invoking a named  method there rather than looking up a hash key); because the `log_if_fail` call will print the entire content of `$body` if it fails here. So just have the message say what it wanted, and the end user can see the full content of the body when that fails.